common/src/main/java/com/microsoft/identity/common/crypto/AndroidAuthSdkStorageEncryptionManager.java

```diff
@@ -0,0 +1,95 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.crypto;
+
+import android.content.Context;
+
+import com.microsoft.identity.common.adal.internal.AuthenticationSettings;
+import com.microsoft.identity.common.java.crypto.StorageEncryptionManager;
+import com.microsoft.identity.common.java.crypto.key.AES256KeyLoader;
+import com.microsoft.identity.common.java.crypto.key.AbstractSecretKeyLoader;
+import com.microsoft.identity.common.java.telemetry.ITelemetryCallback;
+import com.microsoft.identity.common.logging.Logger;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+import lombok.NonNull;
+
+/**
+ * Key Encryption Manager for ADAL & MSAL.
+ * */
+public class AndroidAuthSdkStorageEncryptionManager extends StorageEncryptionManager {
+    private static final String TAG = AndroidAuthSdkStorageEncryptionManager.class.getSimpleName();
+
+    /**
+     * Alias persisting the keypair in AndroidKeyStore.
+     */
+    /* package */ static final String KEY_STORE_ALIAS = "AdalKey";
+
+    private final PredefinedKeyLoader mUserDefinedKey;
+    private final AndroidWrappedKeyLoader mKeyStoreKeyLoader;
+
+    public AndroidAuthSdkStorageEncryptionManager(@NonNull final Context context,
+                                                  @Nullable final ITelemetryCallback telemetryCallback) {
+        if (AuthenticationSettings.INSTANCE.getSecretKeyData() == null) {
+            mUserDefinedKey = null;
+        } else {
+            mUserDefinedKey = new PredefinedKeyLoader("USER_DEFINED_KEY",
+                    AuthenticationSettings.INSTANCE.getSecretKeyData());
+        }
+
+        mKeyStoreKeyLoader = new AndroidWrappedKeyLoader(KEY_STORE_ALIAS, context, telemetryCallback);
+    }
+
+    @Override
+    public @NonNull AES256KeyLoader getKeyLoaderForEncryption() {
+        if (mUserDefinedKey != null) {
+            return mUserDefinedKey;
+        }
+
+        return mKeyStoreKeyLoader;
+    }
+
+    @Override
+    public @NonNull List<AbstractSecretKeyLoader> getKeyLoaderForDecryption(@NonNull byte[] cipherText) {
+        final String methodName = "getKeyLoaderForDecryption";
+
+        if (mUserDefinedKey != null &&
+                isEncryptedByThisKeyIdentifier(cipherText, PredefinedKeyLoader.KEY_IDENTIFIER)) {
+            return new ArrayList<AbstractSecretKeyLoader>() {{
```

Comment:
This [might create a subtle memory leak](https://www.baeldung.com/java-double-brace-initialization#disadvantages-of-using-double-braces), as double-braced initialization keeps a reference to the enclosing class which in this case, has a reference to a `AndroidWrappedKeyLoader` which contains a reference to a `Context`